search2/python/search/utils/beanstalkd.py

```diff
@@ -0,0 +1,199 @@
+            port = publisher_port,
+            queue = event_queue,
+            message = Message(
+                event = progress_event,
+                queueID = job_id,
+                submissionID = submission_id,
+                data = None,
+            ))
+
+def listen(
+    handler_fn: Callable,
+    submit_msg_fn: Callable,
+    completed_msg_fn: Callable,
+    queue_conf: QueueConf,
+    tube: str,
+):
+    """Listen on a Beanstalkd channel, waiting for work.
+       When work is available call the work handler
+    """
+    hosts, ports = queue_conf.split_host_port()
+
+    for event in ("progress", "failed", "started", "completed"):
+        assert event in queue_conf.events
+
+    tube_conf = queue_conf.tubes[tube]
+    events_conf = queue_conf.events
+    clients = tuple(BeanstalkClient(host, port, socket_timeout=10) for (host, port) in zip(hosts, ports))
+
+    i = 0
+    while True:
+        i += 1
+        offset = i % len(hosts)
+        client  = clients[offset]
+
+        client.watch(tube_conf["submission"])
+
+        try:
+            job = client.reserve_job(5)
+        except BeanstalkError as err:
+            if err.message == BEANSTALK_ERR_TIMEOUT:
+                continue
+            raise err
+        try:
+            job_data = loads(job.job_data)
+
+            publisher: Publisher = _specify_publisher(
+                job_data['submissionID'],
+                job.job_id,
+                publisher_host=client.host,
+                publisher_port=client.port,
+                progress_event=events_conf["progress"],
+                event_queue=tube_conf["events"]
+            )
+
+            base_msg = {
+                "event": events_conf["started"],
+                "queueID": job.job_id,
+                "submissionID": job_data["submissionID"],
+            }
+
+            msg = submit_msg_fn(base_msg, job_data)
+
+            client.put_job_into(tube_conf["events"], dumps(msg))
+            res = asyncio.get_event_loop().run_until_complete(handler_fn(publisher, job_data))
+
+            base_msg["event"] = events_conf["completed"]
+            msg = completed_msg_fn(base_msg, job_data, res)
+
+            client.put_job_into(tube_conf["events"], dumps(msg))
+            client.delete_job(job.job_id)
+        except BeanstalkError as err:
+            print(f"Received error during execution: {err}")
+            client.release_job(job.job_id)
+        except NotFoundError as err:
+            msg = {
+                "event": events_conf["failed"],
+                "reason": "404",
+                "queueID": job.job_id,
+                "submissionID": job_data["submissionID"],
+            }
+
+            traceback.print_exc()
+            client.put_job_into(tube_conf["events"], dumps(msg))
+            client.delete_job(job.job_id)
+        except Exception as err:
+            traceback.print_exc()
+            client.release_job(job.job_id)
+        finally:
+            time.sleep(0.5)
```

Comment:
as a general comment about json parsing, if we're sleeping at the end of this block we probably shouldn't worry too much about perf :)

orjson has the advantages of speed and the ability to handle more complex python datatypes, but it has the disadvantage of way worse error messages.  We're not operating in a regime where the speedup is going to be detectable (parsing is, generously, six orders of magnitude under the sleep time) and we're not passing tricky python objects back and forth, but we are going to have to deal with the error messages :)